compiler/rustc_target/src/spec/targets/riscv32_wrs_vxworks.rs

```diff
@@ -0,0 +1,24 @@
+use crate::spec::{base, StackProbeType, Target, TargetOptions};
+
+pub(crate) fn target() -> Target {
+    Target {
+        llvm_target: "riscv32".into(),
+        metadata: crate::spec::TargetMetadata {
+            description: None,
+            tier: Some(3),
+            host_tools: Some(false),
+            std: None, // ?
+        },
+        pointer_width: 32,
+        data_layout: "e-m:e-p:32:32-i64:64-n32-S128".into(),
+        arch: "riscv32".into(),
+        options: TargetOptions {
+            cpu: "generic-rv32".into(),
+            llvm_abiname: "ilp32d".into(),
+            max_atomic_width: Some(32),
+            features: "+m,+a,+f,+d,+c,+zicsr".into(),
```

Comment:
On clang there is the `-print-enabled-extensions` option. When I pass that with `-mcpu=generic-rv32` which is defined by the RISC-V LLVM backend, this is the output:
```
Extensions enabled for the given RISC-V target

    Name                 Version   Description
    i                    2.1       'I' (Base Integer Instruction Set)
    m                    2.0       'M' (Integer Multiplication and Division)
    a                    2.1       'A' (Atomic Instructions)
    f                    2.2       'F' (Single-Precision Floating-Point)
    d                    2.2       'D' (Double-Precision Floating-Point)
    c                    2.0       'C' (Compressed Instructions)
    zicsr                2.0       'zicsr' (CSRs)
    zifencei             2.0       'Zifencei' (fence.i)
    zmmul                1.0       'Zmmul' (Integer Multiplication)
```
Are we missing features here?

Reply:
I think these would be enough. Vxworks compiler driver `wr-cc` would also add any missing arguments.

Reply:
According to [TargetOptions](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/spec/struct.TargetOptions.html#structfield.cpu), 

> cpu: [Cow](https://doc.rust-lang.org/nightly/alloc/borrow/enum.Cow.html)<'static, [str](https://doc.rust-lang.org/nightly/std/primitive.str.html)>
Default CPU to pass to LLVM. Corresponds to llc -mcpu=$cpu. Defaults to “generic”.

> features: [Cow](https://doc.rust-lang.org/nightly/alloc/borrow/enum.Cow.html)<'static, [str](https://doc.rust-lang.org/nightly/std/primitive.str.html)>

Since we are passing `generic-rv32` as cpu, and that automatically enables `i, m, a, f, d, c, zicsr, zifencei, and zmmul` (as per the table above), I am not sure that passing any extra features are needed. I think you can drop the `features` attribute here.
Default target features to pass to LLVM. These features overwrite -Ctarget-cpu but can be overwritten with -Ctarget-features. Corresponds to llc -mattr=$features. Note that these are LLVM feature names, not Rust feature names!

Reply:
Ok sure.

Reply:
If I remove this I get these warnings during compilation. 
```
Hard-float 'd' ABI can't be used for a target that doesn't support the D instruction set extension (ignoring target-abi)
   Compiling getopts v0.2.21
Hard-float 'd' ABI can't be used for a target that doesn't support the D instruction set extension (ignoring target-abi)
Hard-float 'd' ABI can't be used for a target that doesn't support the D instruction set extension (ignoring target-abi)
Hard-float 'd' ABI can't be used for a target that doesn't support the D instruction set extension (ignoring target-abi)
   Compiling test v0.0.0 (/ala-salim-lx12/abbas/rust/nightly/rustLangRust/library/test)
Hard-float 'd' ABI can't be used for a target that doesn't support the D instruction set extension (ignoring target-abi)

```

Reply:
Although the compiler would adds target features for specified cpu, I think these serve as metadata and are helpful. The other target_spec files also specify some features.

Reply:
I think I should add atleast "m, a, f, d, c" as target features.

Reply:
Thanks for giving this a try. Definitely can’t hurt to pass redundant info to LLVM and as you point out, it suppresses some warnings. I wonder if the warnings are coming from rust and not LLVM.

Reply:
I think it's from rust and not llvm. 
Maybe here: https://github.com/rust-lang/rust/blob/1b5aa96d6016bafe50e071b45d4d2e3c90fd766f/compiler/rustc_target/src/target_features.rs#L518